cpp/include/raft/linalg/mdarray.h

```diff
@@ -0,0 +1,389 @@
+
+ private:
+  template <typename E>
+  using view_type_impl = std::conditional_t<
+    container_policy_type::is_host_type::value,
+    host_mdspan<E, extents_type, layout_type, typename container_policy_type::accessor_policy>,
+    device_mdspan<E, extents_type, layout_type, typename container_policy_type::accessor_policy>>;
+
+
+ public:
+  /**
+   * \brief the mdspan type returned by view method.
+   */
+  using view_type = view_type_impl<element_type>;
+  using const_view_type = view_type_impl<element_type const>;
+
+ public:
+  constexpr mdarray() noexcept(std::is_nothrow_default_constructible<container_type>::value) =
+    default;
+  constexpr mdarray(mdarray const&) noexcept(
+    std::is_nothrow_copy_constructible<container_type>::value) = default;
+  constexpr mdarray(mdarray&&) noexcept(std::is_nothrow_move_constructible<container_type>::value) =
+    default;
+
+  auto operator =(mdarray&&) noexcept(std::is_nothrow_move_assignable<container_type>::value)
+    -> mdarray& = default;
+
+  auto operator =(mdarray const&) noexcept(std::is_nothrow_copy_assignable<container_type>::value)
+    -> mdarray& = default;
+
+  ~mdarray() noexcept(std::is_nothrow_destructible<container_type>::value) = default;
+
+  /**
+   * @brief The only constructor that can create storage, this is to make sure CUDA stream is being
+   * used.
+   */
+  constexpr mdarray(mapping_type const& m, container_policy_type const& cp)
+    : cp_(cp), map_(m), c_(cp_.create(map_.required_span_size()))
+  {
+  }
+  /**
+   * @brief Get a mdspan that can be passed down to CUDA kernels.
+   */
+  auto view() noexcept { return view_type(c_.data(), map_, cp_.view()); }
+  /**
+   * @brief Get a mdspan that can be passed down to CUDA kernels.
+   */
+  auto view() const noexcept { return const_view_type(c_.data(), map_, cp_.view()); }
+
+  auto data() noexcept -> pointer { return cp_.data(); }
+  constexpr auto data() const noexcept -> const_pointer { return cp_.data(); }
+
+  /**
+   * @brief Indexing operator, use it sparingly since it triggers a device<->host copy.
+   */
+  template <typename... IndexType>
+  auto operator()(IndexType&&... indices)
+    -> std::enable_if_t<sizeof...(IndexType) == extents_type::rank() &&
+                          (std::is_convertible_v<IndexType, index_type> && ...) &&
+                          std::is_constructible_v<extents_type, IndexType...> &&
+                          std::is_constructible_v<mapping_type, extents_type>,
+                        /* device policy is not default constructible due to requirement for CUDA
+                           stream. */
+                        /* std::is_default_constructible_v<container_policy_type> */
+                        reference>
+  {
+    return cp_.access(c_, map_(std::forward<IndexType>(indices)...));
+  }
+
+  /**
+   * @brief Indexing operator, use it sparingly since it triggers a device<->host copy.
+   */
+  template <typename... IndexType>
```

Comment:
Does it make sense to throw a compiler warning when this function is used?

Reply:
It won't be possible to determine whether it's a device array at compile time

Reply:
We can limit it to debug build if necessary.